modules/auxiliary/server/capture/ldap.rb

```diff
@@ -0,0 +1,181 @@
+                  end
+                elsif user_login.name =~ /,/
+                  begin
+                    dn = Net::LDAP::DN.new(user_login.name)
+                    dn.each_pair do |key, value|
+                      if key == 'cn'
+                        names << value
+                      elsif key == 'dc'
+                        domains << value
+                      end
+                    end
+                    @state[client][:user] = names.first
+                    @state[client][:domain] = domains.empty? ? nil : domains.join('.')
+                  rescue StandardError => e
+                    print_error("LDAP Login Attempt => From:#{@state[client][:name]} DN:#{user_login.name}")
+                    raise e
+                  end
+                else
+                  result_code = Net::LDAP::ResultCodeInvalidCredentials
+                end
+              else
+                @state[client][:user] = ''
+                @state[client][:domain] = nil
+              end
+
+              @state[client][:pass] = user_login.authentication
+
+              unless @state[client][:user].empty? && @state[client][:pass].empty?
+                report_cred(
+                  ip: @state[client][:ip],
+                  port: client.localport,
+                  service_name: 'ldap',
+                  user: @state[client][:user],
+                  password: @state[client][:pass],
+                  domain: @state[client][:domain]
+                )
+              end
+              print_good("LDAP Login Attempt => From:#{@state[client][:name]} Username:#{@state[client][:user]} Password:#{@state[client][:password]}")
+              result_code = Net::LDAP::ResultCodeAuthMethodNotSupported if result_code.nil?
+              service.encode_ldap_response(
+                pdu.message_id,
+                result_code,
+                '',
+                'Try Again or Use a different Authentication Method',
+                Net::LDAP::PDU::BindResult
+              )
+            else
+              result_code = Net::LDAP::ResultCodeUnwillingToPerform
+              service.encode_ldap_response(
+                pdu.message_id,
+                result_code,
+                '',
+                Net::LDAP::ResultStrings[result_code],
+                Net::LDAP::PDU::SearchResult
+              )
+            end
+
+      res.nil? ? client.close : on_send_response(client, res)
+    rescue StandardError => e
+      on_send_response(client,
+                       service.encode_ldap_response(
+                         1,
+                         Net::LDAP::ResultCodeUnwillingToPerform,
+                         '',
+                         "Failed to handle LDAP request due to #{e}",
+                         Net::LDAP::PDU::BindResult
+                       ))
+      print_error("Failed to handle LDAP request due to #{e}")
+      client.close
+    end
```

Comment:
Should completion of the request alway `close` the client?  If so consider putting it in an `ensure` for the method call.

```suggestion
    end
    ensure
      client.close
```

Reply:
Is it appropriate now?